### PR TITLE
Meghana reddy step 17

### PR DESCRIPTION
**Pull request by Meghana-Reddy2**
DESCRIPTION 

Step 17: Implement login/logout functionality
Try implementing it yourself without using additional gems
The purpose is to deepen your understanding of mechanisms such as HTTP cookies and Sessions in Rails by not using gems such as Devise.
It also aims to deepen understanding of general authentication (e.g., how to handle passwords).
Let's implement the login screen
If you are not logged in, make sure you are not able to transition to the task management page.
Show only tasks you've created
Let's implement the logout function
Screenshot 2024-03-12 at 15 56 06
Screenshot 2024-03-12 at 15 56 23
Screenshot 2024-03-12 at 15 56 33
Screenshot 2024-03-12 at 15 59 25

### CODE REVIEW

File: myapp/app/controllers/application_controller.rb
Comment (wataru0) [inline myapp/app/controllers/application_controller.rb:1]: There are indentation mis-matching

File: myapp/app/assets/stylesheets/session.scss
Comment (wataru0) [inline myapp/app/assets/stylesheets/session.scss:1]: Is this file needed?

File: myapp/app/models/task.rb
Comment (wataru0) [inline myapp/app/models/task.rb:1]: I think there are commit failing. Some parts are duplicated.

File: myapp/db/seeds.rb
Comment (wataru0) [inline myapp/db/seeds.rb:1]: I think this is not working. there are some duplications. please check `bundle exec rails db:seed`.

**Review by wataru0 — COMMENTED** [review]
I think there are some un-needed files. please check auto-created files 👍

File: myapp/app/assets/stylesheets/session.scss
Comment (Meghana-Reddy2) [reply myapp/app/assets/stylesheets/session.scss:1]: No, This was automatically generated. I didn't use this in my designing. So I will just delete it. 🙇

File: myapp/db/seeds.rb
Comment (Meghana-Reddy2) [reply myapp/db/seeds.rb:1]: I didn't get that error when I first ran the rails db:seed command. As I have a uniqueness constraint on my username running rails db:seed again is showing an error as the username already exists 👀

File: myapp/app/controllers/application_controller.rb
Comment (Meghana-Reddy2) [reply myapp/app/controllers/application_controller.rb:1]: Corrected it. These must have been generated after I ran rubocop and due to some merge conflicts 😢

File: myapp/app/models/task.rb
Comment (Meghana-Reddy2) [reply myapp/app/models/task.rb:1]: Deleted it. This was due to merge commits

File: myapp/db/seeds.rb
Comment (wataru0) [reply myapp/db/seeds.rb:1]: Yep, fixed by https://github.com/Fablic/training/pull/1620/commits/647224e8b29b7a03e87885abea5b8d9108c2cece 👍

File: myapp/app/helpers/session_helper.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    return unless session[:user_id]
+    @current_user ||= User.find_by(id: session[:user_id])
+  end
+    
+  def current_user?(user)
+    user == current_user
+  end
+    
+  def logged_in?
+    !current_user.nil?
```
Comment (wataru0) [inline myapp/app/helpers/session_helper.rb]:
```suggestion
    current_user.present?
```
how about this? Does this work well?

File: myapp/app/views/session/create.html.erb
Comment (wataru0) [inline myapp/app/views/session/create.html.erb:1]: this view file won't be displayed. so can we remove it?

File: myapp/app/views/session/destroy.html.erb
Comment (wataru0) [inline myapp/app/views/session/destroy.html.erb:1]: This won't be displayed as well. so remove it!

File: myapp/config/locales/en.yml [hunk 2]
```diff
@@ -53,12 +53,13 @@ en:
       not started: "Not Started"
       in progress: "In Progress"
       done: "Done"
-
+
```
Comment (wataru0) [inline myapp/config/locales/en.yml]: Is this difference intended? if no, you can revert. 👀

File: myapp/config/locales/ja.yml [hunk 3]
```diff
@@ -26,7 +27,7 @@ ja:
     status:
       not started: "始まっていない"
       in progress: "進行中"
-      done: "終わり"
+      done: "終わり"
```
Comment (wataru0) [inline myapp/config/locales/ja.yml]: This whitespace is also intended? 👀

File: myapp/spec/helpers/session_helper_spec.rb
Comment (wataru0) [inline myapp/spec/helpers/session_helper_spec.rb:1]: If you don't have a plan to test this function, you should delete this file.

File: myapp/spec/requests/session_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    let (:req_params) { { session: { username: 'testUser', password: 'password' } } }
+
+    it 'returns http success', :aggregate_failures do
+      post '/login', params: req_params
+      expect(response).to have_http_status(:found)
+      if I18n.locale.to_s == "en"
+        expect(response).to redirect_to('/tasks?locale=en')
+      elsif I18n.locale.to_s == "ja"
+        expect(response).to redirect_to('/tasks?locale=ja')
+      end
```
Comment (wataru0) [inline myapp/spec/requests/session_spec.rb]: I think these are not good for testing. 

Using if statements can make the behavior of tests dependent on external states or conditions. As a result, the same test could potentially produce different outcomes under different circumstances.

In general, it's recommended to use context or describe blocks to separate test cases, making sure each test case clearly tests one scenario or state.

File: myapp/spec/views/session/new.html.erb_spec.rb
Comment (wataru0) [inline myapp/spec/views/session/new.html.erb_spec.rb:1]: what do you want to test in this file? 👀

**Review by wataru0 — COMMENTED** [review]
some comments 👍

File: myapp/app/views/session/create.html.erb
Comment (Meghana-Reddy2) [reply myapp/app/views/session/create.html.erb:1]: Yes. Deleted it

File: myapp/config/locales/en.yml (hunk 2 above)
Comment (Meghana-Reddy2) [reply myapp/config/locales/en.yml]: Reverted it back.

File: myapp/config/locales/ja.yml (hunk 3 above)
Comment (Meghana-Reddy2) [reply myapp/config/locales/ja.yml]: Removed it 😥

File: myapp/spec/views/session/new.html.erb_spec.rb
Comment (Meghana-Reddy2) [reply myapp/spec/views/session/new.html.erb_spec.rb:1]: Just the Login functionality 👀

File: myapp/spec/requests/session_spec.rb (hunk 4 above)
Comment (Meghana-Reddy2) [reply myapp/spec/requests/session_spec.rb]: Updated the modifications plz check [9b2f479](https://github.com/Fablic/training/pull/1620/commits/9b2f479bffbe858a7027925a30f4b2fb2cdfba69)

File: myapp/spec/views/session/new.html.erb_spec.rb
Comment (wataru0) [inline myapp/spec/views/session/new.html.erb_spec.rb:1]: Screenshot 2024-03-14 at 18 12 30
In my environment, this test is failed.

Comment (wataru0) [reply myapp/spec/views/session/new.html.erb_spec.rb:1]: Screenshot 2024-03-14 at 18 12 30
In my environment, this test is failed.

Comment (Meghana-Reddy2) [reply myapp/spec/views/session/new.html.erb_spec.rb:1]: Sorry , This is not for testing. I confused it with my new.html.erb views file and thought about my session_spec test when u asked about testing 😥 . This is an extra file which should be removed. I forgot to delete that file. I have deleted it please check it once. Sorry for the confusion. 🙇‍♀️

File: myapp/spec/system/tasks_spec.rb [hunk 5]
```diff
@@ -1,27 +1,47 @@
...
+
 RSpec.describe 'Tasks', type: :system do
+
   describe 'page rendering' do
+    before do
+      create_user
+      login
+    end
+
+    let!(:user) { User.last }
```
Comment (wataru0) [inline myapp/spec/system/tasks_spec.rb]: I think if there is no reason to get last user, looks good to me as follows
```suggestion
    let!(:user) { create(:user) }
```
This is because it is unpredictable which user will be the last, depending on the state of the database and the order of testing.
Tests should be unaffected by the state of the database or the order of the tests and should have consistent results.

File: myapp/spec/system/tasks_spec.rb [hunk 6]
```diff
@@ -34,6 +54,68 @@
 
   describe 'task creation' do
     before do
+      create_user
+      login
+    end  
+    it 'task created successfully & show flash message when task created', :aggregate_failure do
+        visit '/'
+        click_link('New task')
+        # expect(Task.all.length).to eq 0
```
Comment (wataru0) [inline myapp/spec/system/tasks_spec.rb]: if no reason to put this comment for another developer, it's better to remove it 👍

File: myapp/spec/system/tasks_spec.rb [hunk 7]
```diff
@@ -34,6 +54,68 @@
...
+        fill_in 'task[name]', with: 'a_new_task'
+        fill_in 'task[description]', with: 'new task description'
+        select 'Done', from: 'Status'
+        select 'Low', from: 'Priority'
+        fill_in 'task[duedate]', with: '2024-08-08'
+        click_on "Update"
+        expect(page).to have_content "a_new_task"
+        expect(page).to have_link href: /\/tasks\/\d+/
+        expect(Task.last.name).to eq 'a_new_task'
+        # expect(Task.all.length).to eq 1
```
Comment (wataru0) [inline myapp/spec/system/tasks_spec.rb]: same

File: myapp/spec/system/tasks_spec.rb [hunk 8]
```diff
@@ -34,6 +54,68 @@
...
+  
+      it 'task updated successfully' do
+        visit '/'
+        click_on('Edit')
+        fill_in 'task[name]', with: 'task_after_edit'
+        fill_in 'task[description]', with: 'description after'
+        click_on "Update"
+        expect(page).to have_content 'task_after_edit'
+        expect(Task.last.name).to eq 'task_after_edit'
+        # expect(Task.all.length).to eq 1
```
Comment (wataru0) [inline myapp/spec/system/tasks_spec.rb]: same

File: myapp/spec/system/tasks_spec.rb [hunk 9]
```diff
@@ -34,6 +54,68 @@
...
+    describe 'task deletion' do
+      before do
+        user = create_user
+        create(:task, name: 'task_should_be_deleted', user: user)
+        login
+      end
+  
+      it 'task deleted successfully & show flash message when task deleted', :aggregate_failures do
+        visit '/'
+        # expect(Task.all.length).to eq 1
```
Comment (wataru0) [inline myapp/spec/system/tasks_spec.rb]: same

File: myapp/spec/system/tasks_spec.rb [hunk 10]
```diff
@@ -34,6 +54,68 @@
...
+        click_on "Delete"
+        # expect(Task.all.length).to eq 0
```
Comment (wataru0) [inline myapp/spec/system/tasks_spec.rb]: same

File: myapp/app/helpers/session_helper.rb (hunk 1 above)
Comment (Meghana-Reddy2) [reply myapp/app/helpers/session_helper.rb]: Modified this 👍

File: myapp/app/models/task.rb [hunk 11]
```diff
@@ -1,17 +1,20 @@
...
+
   belongs_to :user
-  STATUSES = ['Not Started', 'In Progress', 'Done'].freeze
-  PRIORITY = ['High', 'Medium', 'Low'].freeze
-  validates :name, presence: true, length: { maximum: 30, length: { maximum: 30, long: I18n.t('flash_msgs.long') } }
-  validates :description, length: { maximum: 255, length: { maximum: 255, long: I18n.t('flash_msgs.long') } }
+  STATUSES = ["Not Started", "In Progress", "Done"].freeze
+  PRIORITY = ["High" , "Medium", "Low"].freeze
+  validates :name, presence: true ,length: {maximum: 30, length: { maximum: 30, long: I18n.t('flash_msgs.long') } }
+  validates :description , length: {maximum: 255, length:{ maximum: 255, long: I18n.t('flash_msgs.long')} }
```
Comment (wataru0) [inline myapp/app/models/task.rb]: please check and fix these differences 👀

**Review by wataru0 — COMMENTED** [review]
good job 👍 left comments 🙏

**Review by wataru0 — APPROVED** [review]
LGTM
